### ISSUE TITLE
Alert notification remind frequency

### ISSUE THREAD

**Issue by kapustor** [grafana/grafana#7330]
Hi.
Now alerts send notification for two times: one for "Alerting", one for "OK". It would be great to have a possibility to setup additional notification (both slack and email) while alert is still active and triggering every N hours (10h, 10m, etc). It can be an option when creating alert.

Thank you.

**Comment by DerekL** [comment]
+1

**Comment by yannispanousis** [comment]
100%

**Comment by rmrf** [comment]
This is a *must* feature for any Alerting system. need this ASAP.

**Comment by dirkaholic** [comment]
Would definitely improve the alerting system especially when it would be possible to configure that on alert level.

**Comment by andreioss** [comment]
+1!
Just a necessary feature!

**Comment by swift1911** [comment]
+1
Very necessary

**Comment by ejurgelionis** [comment]
+1

**Comment by theobolo** [comment]
+1 Absolutely and definitely essential !

**Comment by gloine** [comment]
+1

**Comment by bergquist** [comment]
Agree. We will try to get this into 5.0

**Comment by bergquist** [comment]
My suggested solution for this is to introduce a new table (alert_notification_journal) for sent notifications so we can keep track of the last sent notification per alert and notifier. The table should look something like this: `id, alert_id, notifier_id, sent_at, succeed`

Before sending the notification, we should read the history for the alert and pass the history to each notifier. 
Each notifier should then be able to decide if it should send a new notification. We should provide a base implementation that most notifiers can default to. But some notifiers always want to send notifications so they should be able to do so and just ignore the history. 

We should also introduce a cleanup job for this. So we only keep history for N days or something like that. 

It should also be possible to configure a min interval for each notifier in the notifiers edit page. Ex `10m`

**Comment by phateuk** [comment]
+1million. This would be amazing

**Comment by jbaublitz** [comment]
@torkelo Not sure if you're looking for a PR for this but I'm happy to help out if you're interested. Shoot me an email or @ me here if you'd like to coordinate.

**Comment by bergquist** [comment]
@jbaublitz A PR would be appreciated. Sounds like there might be some started work in another issue. But no PR's yet. https://github.com/grafana/grafana/issues/7481#issuecomment-350446228

**Comment by jbaublitz** [comment]
@bergquist Excellent! Would you have some time to chat on IRC about the details? I have a basic idea of what I would like to do. The short of it is a few changes to `EvalContext` so that it keeps track of the number of evaluations while `Firing` is true. Once it hits the specified number of evaluations, it can reset and notify. Something along those lines. I'm assuming this will require some model changes as well and I'd love to touch base with you about your thoughts around this before I submit a PR.

**Comment by bergquist** [comment]
@jbaublitz I would really prefer to keep the discussion here. Makes it easier to track and make it possible for others to give feedback. 

I don't think the EvalContext should care about how often a notification should be sent or what notifiers should be sending something. I would rather prefer each notifier keeping its own state stored in the database (log of sent notifications is a good enough state in this case).

**Comment by jbaublitz** [comment]
@bergquist Okay not a problem. Happy to discuss here. I looked into the code more after posting that and realized that your suggestion makes more sense so I'll move forward with that. I'm currently setting up a development environment and will post back here if I run into any issues. Thanks.

**Comment by sushovan23** [comment]
would this feature reevaluate the panel? so this would somehow resolve the issue of multiple series not sending alerts when panel is already in firing state?

**Comment by jbaublitz** [comment]
@bergquist I have the backend code in place it seems but am not a frontend person and am struggling with the UI modifications. How should I proceed?

**Comment by jbaublitz** [comment]
@bergquist I did some looking around the code base and was able to make the changes to the UI. Will be testing both the UI and API over the next couple of days and will hopefully have a PR for you by the end of next week. I'll keep you posted.

**Comment by jbaublitz** [comment]
@phateuk Happy to help!

For those following along, I got the old version of master working with repeated alerting per number of evaluations. However, some of the alerting seems to have changed in a way that breaks automerge/rebase so I need to take a look at how to merge manually - doing that and some additional testing today. For now, here is a [commit](https://github.com/jbaublitz/grafana/commit/27f1b110eed16129a0835f777a627561ee84ae6f) that yields a working version of the behavior described above.

**Comment by romankor** [comment]
+1

**Comment by jsedy7** [comment]
+1

**Comment by AODBA** [comment]
+1

**Comment by chesterlai** [comment]
+1

**Comment by bashartman** [comment]
+1

**Comment by hjyun328** [comment]
+1

**Comment by theobolo** [comment]
+1

**Comment by osimar-medeiros** [comment]
+1

**Comment by kolmio23** [comment]
+1

**Comment by gaubapravar** [comment]
Is there any update on this @jbaublitz @bergquist @kapustor . I am facing the similar issue! I want that the alerts should be triggered after every N minutes. But Grafana(by default) triggers only twice. Stuck in this!

**Comment by vainkop** [comment]
+1

**Comment by jbaublitz** [comment]
@gaubapravar hey there - I can follow up with Grafana folks on my PR for this issue.

Not sure what's going on on the Grafana side but to anyone tempted to post another +1, I have a pull request waiting on a response and I don't think that adding to all of the +1's will get it merged any faster. Based on the thumbs down reactions, I don't think the people being notified when the +1's are posted are particularly happy about this. Practically it's probably best to wait until we hear back for now.

**Comment by jbaublitz** [comment]
@bergquist From the link above:
> My current thinking is that we would have to implement #7330 first so it would be configurable how frequently anyone wants the evalMatches changes to be pushed.

Just want to get a feel for if by "implement" you mean the core Grafana team would prefer to do this vs. taking an external patch. I already have a PR ready for review that provides this feature and am happy to edit it if need be if my code doesn't meet the requirements that you have for this issue. I mainly just want to know if I should keep updating the patch or if my pull request is eventually going to get closed. Please let me know when you get the chance - I am happy to work with you and will gladly take feedback on the PR if changes need to be made.

**Comment by bergquist** [comment]
First of all. Thank you for your patience @jbaublitz 

I've been thinking about this feature and the way I would like to see this feature is to move the reminder setting from the alert rule to the alert notification channel. I can't really see any good reason for configuring different reminder per alert rule. Perhaps someone else can give a good example of when this is needed? Keeping a log of outgoing alert notifications would also be more resilient since failed notifications could be retried. Any reason for not using an `alert_notification_journal` as I suggested in my previous comment?

**Comment by meeki007** [comment]
+1

**Comment by jbaublitz** [comment]
@meeki007 I got a response on PR #10528 so the conversation has moved there. I'm working on getting the changes requested ready for review. Feel free to follow along on the PR if you'd like to see what's going on. Now that we've agreed on design decisions, I'm actively making the changes so this should be ready for a second round of review soon.

**Comment by alexandersauerbier** [comment]
+1

**Comment by pete00067** [comment]
Now grafana has a send reminders option now in the Alert Notifications page. Which allows notification to be sent every interval(selected by you) while alert is still on.

**Comment by jianfengye** [comment]
Is it grafana can send alert every interval in grafana7.3.5?

**Comment by jianfengye** [comment]
> Is it grafana can send alert every interval in grafana7.3.5?

i found that in notification-> reminder, thanks

**Comment by qixiaobo** [comment]
> > Is it grafana can send alert every interval in grafana7.3.5?
> 
> i found that in notification-> reminder, thanks

alerting /notification/Notification settings
![image](https://user-images.githubusercontent.com/2881751/123728006-2cff0700-d8c5-11eb-8ef1-0571af305922.png)

**Comment by Smasherr** [comment]
This has been moved to "old alerting" now. The "new alerting" seems to send reminders every 4 hours and I cannot find a possibility to disable it.

**Comment by daniel-salmon** [comment]
@Smasherr , if you go under Alerting --> Notification Policies there is a "Repeat Interval" setting for each policy (you can set it for the root policy or override it for each sub-policy). There doesn't seem to be a way to disable it (that I can see), but you can set the interval to an absurdly long time duration as a work around....
